### ISSUE TITLE
regarding the error msg

### ISSUE THREAD

**Issue by ctf0** [codezero-be/laravel-unique-translation#1]
atm to get the error you will use `name.code` but is there a way to pass an option that make the error msg unified whether am using a code or not ?

ex.
```php
// form
<input name="slug[en]" value="abc">
<input name="slug[nl]" value="abc">

// validate
$attributes = request()->validate([
    'slug.*' => ["unique_translation:posts,slug,{$post->id}"],
]);

// error
@if($errors->has('slug'))
    {{ $errors->first('slug') }}
@endif
```

this way , either i use a single input or array, the error condition doesnt change.

atm, here is how i do the validation
```php
foreach ($locales as $code) {
    $v = Validator::make($request->all(), ["slug.$code" => "unique:pages,slug->$code," . $id ?: '']);

    if ($v->fails()) {
        $validator->errors()->add('slug', 'this slug is already taken');
    }
}
```

### on a side note

also is it possible to use the validation rule with `|` pipe rules or do i have to use the array instead ?

**Comment by ivanvermeyen** [comment]
Hello

If I understand it correctly your problem is that the error key is different when you:

- validate a single translation:

```php
$attributes = request()->validate([
    'slug' => ["unique_translation:posts,slug,{$post->id}"],
]);

$errors->first('slug'); // <= does not include the locale
```

- vs an array of translations:

```php
$attributes = request()->validate([
    'slug.*' => ["unique_translation:posts,slug,{$post->id}"],
]);

$errors->first('slug.en'); // <= includes the locale
```

At the moment this is [basic Laravel behavior](https://laravel.com/docs/5.5/validation#working-with-error-messages). I just provide the error message, but Laravel adds it to the session based on the attribute key in your validation rules.

I see how it could be useful to always add the `slug.locale` (and `slug`?) error, but then I would need a reference to the validator instance. I have this in the "old" validator class, but not in the Laravel 5.5 Rule. I will need to dig a bit deeper into this to see if it's possible.

You should be able to use the `|` symbol as with any Laravel validation rule.

**Comment by ctf0** [comment]
okay np, thanx.

**Comment by ivanvermeyen** [comment]
It would be cool if Laravel would pass the Validator instance to the Rule, but it doesn't :/

```php
// \Illuminate\Validation\Validator::validateUsingCustomRule
protected function validateUsingCustomRule($attribute, $value, $rule)
{
    if (! $rule->passes($attribute, $value)) {
        $this->failedRules[$attribute][get_class($rule)] = [];

        $this->messages->add($attribute, $this->makeReplacements(
            $rule->message(), $attribute, get_class($rule), []
        ));
    }
}
```

I did find this article that shows an alternate approach for what you were doing, using an `after` callback on the `Validator`. Maybe you can create a `BaseFormRequest` that adds the errors on some condition (you can get an array of failed attributes from the Validator), and then all of your `FormRequest` classes can extend it? It's just a wild idea.

https://pascalbaljetmedia.com/nl/blog/working-with-the-validator-instance-in-form-requests

**Comment by ctf0** [comment]
the one from the article is related to the formrequest class, which i already aware of and even have a away around for the multi file upload which laravel didnt fix since 4.2 "which is big focken idiotic 😠 ",

and even with the new `withValidator` its not quite helpful in its current form because what if the user want to validate something form the request ?

still here is the article if u r interested https://ctf0.wordpress.com/2016/10/16/extend-formrequest-to-allow-more-functionality-in-laravel-v5-3/

about getting the validating inst, actually laravel support that https://laravel.com/docs/5.5/validation#using-extensions

other than that, i do know that there IS a way around but am a bit busy with a couple of other things atm to dig deeper into this.

maybe as temp solution for now, we can use it along with the old way of `Validator::make`

**Comment by ivanvermeyen** [comment]
Hmm, I currently have 2 classes:

- the Validator extension `UniqueTranslationValidator`
- the 5.5 Rule `UniqueTranslationRule`

The Validator extension uses the Rule's logic to perform the validation.

I think I can turn this around if I make the Rule behave like one of Laravel's [Rule classes](https://github.com/illuminate/validation/blob/master/Rules/Unique.php). These just return the string representation of the validation rule. This would allow access to Laravel's `Validator` instance to set custom errors...

I'll try and refactor this! :)

What are your thoughts on this: should there always be a `slug.locale` error AND `slug` error?
Or should I make it an option somehow to "force" both? I don't know if there are use cases where this behavior would be undesirable?

**Comment by ctf0** [comment]
i think u should keep both, why ?

because 99& of users are accustomed to what the official docs says, but there is also this tiny winy 1% "am one of those" that prefer to keep things simple & easy to use no matter what the underlayer is.

if u want the truth, we could have already solved this by making some array check in the view and display the error accordingly,
but why do that when it's possible to fix this damn shenanigans from the start ?

**Comment by ivanvermeyen** [comment]
Ok, I've pushed a ton of updates. I think I got everything to work properly.
Making it compatible with custom error messages was a bit tricky. 

Here's what I've done:

-   Refactor and merge tests
-   Rewrite Validator extension
-   Update Validation Rule to return a string representation of the rule
-   Always return an error for both the attribute name (`$errors->first('slug')`) and the localized attribute name (`$errors->first('slug.en')`) when validation fails
-   Handle custom error messages
-   Add name Rule constructor: `UniqueTranslationRule::for($table)`

I think this should solve your problem?

Let me know if there any issues with this. 👍
